### PR TITLE
Make "SlugService" overridable by binding it to the container

### PR DESCRIPTION
**Pull request by qodeboy**
As mentioned in #320 we need a way to override `SlugService` implementation to work out cross-model slugs.

Changes outline:
- Add `laravel/framework` as dependency. Not sure it is needed, but `app()` helper is now used across code base.
- Create `SlugContract` interface.
- Make `SlugService` implement it.
- Change all code which used `SlugService` directly to either `SlugContract` typehint or `app(SlugContract::class)` call.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
   ],
   "require": {
     "php": "^5.6 || ^7.0",
+    "laravel/framework": "5.1.* || 5.2.* || 5.3.*",
```
Comment (cviebrock) [inline composer.json]: If we require this, we probably don't need to require the `illuminate/*` packages as well, do we?

Also, does this break support in Lumen?

**Comment by cviebrock** [comment]
Cool ... thanks for this!  Just had a question about the change to `composer.json` (see above).  Otherwise, this looks great.

**Comment by qodeboy** [comment]
That actually a good question, and I don't have an answer off the top of my head. I'll investigate this further and update PR.

I guess we might not need ```laravel/framework``` at all, because this package is designed to be used in laravel ecosystem only, so ```Illuminate/Foundation``` helpers will be always available. Not sure about lumen though.

**Comment by cviebrock** [comment]
_ping_ 😉

**Comment by cviebrock** [comment]
pinging again ...

**Comment by qrazi** [comment]
I am actually migrating an older API from Laravel to Lumen. This package does not seem to be compatible with Lumen currently already, for instance, Lumen by default does not have config_path(). I would however like Lumen compatibility. I think it would also make sense, since Eloquent is officially supported by Lumen?

**Comment by cviebrock** [comment]
This MR is a bit out of date now and my attempt to update the conflicts didn't work.

@qodeboy: I like this idea, so if you have time to resubmit a MR that binds the service to the container, I would gladly accept it.  Thanks!

**Comment by cviebrock** [comment]
@qrazi Can you open another issue for this, since your comment doesn't appear to be related to the original MR.

**Pull request closed without merging**